src/functional/ffront/mockitir.py

```diff
@@ -0,0 +1,215 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from typing import Union
+
+from functional.iterator import ir as itir
+
+
```

Comment:
I think the functions in this module could be added to the `foast_to_itir` module, since they are basically makers to reduce boilerplate code in the creation of iterator IR. If you think they should go in a different module because it is easy to share them for different purposes (like testing), I think the current name is confusing and would suggest something more explicit like `itir_makers`.

Additionally, to make clear that they are makers for itir nodes, I think it'd be worth to use a naming convention like `make_<itir_node>()` or `_<itir_node>()` or `<itir_node>_()` and get rid of the one-line makers for trivial calls to builtin functions, since they are not really needed and in the lowering code it would look like that there are nodes in the iterator IR with those names.

Reply:
I can agree to adding one or maximum two characters to the makers to distinguish them from iterator builtins.
Currently this is achieved by how they are imported (`mi.<itir_node>`).

I can also agree to renaming the module they live in. I do think however that they merit their own module because they are useful beyond the lowering (and even testing).

The following example illustrates why I can not support dropping the builtins shortcuts.

```python
snippet = lambda_("a", "b", "c", ... )(
    plus_(deref_("a"), plus_(deref_("b"), plus_(...)))
)(...)

# vs

snippet = call_("lambda")(
    ["a", "b", "c", ...], 
    call_("plus")(call_("deref")("a"), call_("plus")(call_("deref")("b"), call_("plus")(...))))
)(...)
```

The former allows you to express (**precisely** I might argue) that this snippet creates the IR for a fundef that looks exactly the same, except that variable names wouldn't be strings.
There is a reason why all design notes and discussion notes put field view code next to iterator view fundefs **and not** iterator IR.

The latter is on one hand a more accurate description of the IR nodes, but on the other hand it is also simply a more verbose and harder to read version of the former (do the parentheses actually match? - who can tell without reformatting it or using an IDE?)

Since this seems to become obvious only after trying different ways for 3-4 days maybe it requires an ADR?

Reply:
Ok, I understand your reasons for keeping the one-liners and it's fine for me. It would be enough with adding just a single line/paragraph to the existing ADR mentioning that makers exist for itir symbols to simplify the understanding of the itir expressions (specially in tests, which I think it's the only place where heavily nested code can appear).